### ISSUE TITLE
video diffusion slow down

### ISSUE THREAD

**Issue by eps696** [maua-maua-maua/maua#6]
when processing lengthy (at least few minutes) videos with diffusion, the process gets slowing down progressively (probably taking days instead of hours). it may be due to the big optflow data files (in my case ~5gb); may it be more optimal to save them the per frame?

**Comment by JCBrouwer** [comment]
Hi @eps696 thanks for reporting, I'll admit that I haven't tried videos of much more than 30 seconds.

Do you know specifically which part of the process is getting slowed down?

It should be pretty simple to rewrite the frame loading code to use compressed image files instead of memory-mapped arrays. I figured the memory-mapping would avoid some compute overhead at the cost of extra space, but I guess there are some hidden costs for larger files.

You can give it a try yourself if you want, otherwise I can do it by the end of next weekend.

The responsible class is [here](https://github.com/maua-maua-maua/maua/blob/main/maua/diffusion/video.py#L52). You would only need to implement `__len__`, `__getitem__`, and `append` to load and save image files (`clear`, `update`, `__enter__`, and `__exit__` can all just pass in this case). You'd need to check the shape of the incoming items to determine if they're frames (B,C,H,W), flow (H,W,2), or flow consistency (H,W) and then save/load them to a directory accordingly.

**Comment by eps696** [comment]
yup thanks, i changed it to separate npy files (sample code is below, just in case), will run it again overnight to see if it helps.

```
class FiledFrames(Dataset):
    def __init__(self, file, height, width, device):
        super().__init__()
        self.file, self.height, self.width = file, height, width
        self.device = device
        self.dir = os.path.dirname(self.file)
        self.count = len([f for f in file_list(self.dir, 'npy') if os.path.basename(self.file) in f])

    def __len__(self):
        return self.count

    def __getitem__(self, idx):
        frame = np.load(self.file + '-%05d.npy' % idx)
        tensor = torch.from_numpy(frame).float().to(self.device)
        if tensor.dim() < 4:
            tensor = tensor.unsqueeze(1)
        return tensor

    def append(self, item):
        if isinstance(item, torch.Tensor):
            item = item.detach().cpu().numpy()
        if len(item.shape) < 4:
            item = item[None]
        frame = np.save(self.file + '-%05d.npy' % self.count, item)
        self.count += 1

```

**Comment by eps696** [comment]
well, no need for overnight run - this time it's rendered within 3 hours, keeping the same speed down to milliseconds.
so the problem was indeed in the huge single-file arrays.

**Comment by JCBrouwer** [comment]
Cool, looks like great solution! I'll look at integrating something similar into the main script. I think it's worth it to do the shape checks I mentioned above and then save them as image files rather than .npy files. That way people can see the progress of videos on their filesystem.

I'll go out on a limb and say that NpyAppendArray is probably seek()ing the entire file with each append which makes the operation linear in number of frames. This'll get out of hand pretty quick for larger videos.

**Comment by eps696** [comment]
i did it first, but the flow images were not that informative, and png files are waay slower to save than npy. that could be sorted with paralleling saving process of course, yet i decided to stay with arrays. i moved also from keeping the whole output in the memory, saving frames as separate jpegs [with quality 95% the loss is neglectable] and then converting them to video with a single external ffmpeg command.
as for NpyAppendArray - it's still surprising that seeking within a file is so different perf-wise than seeking file in the directory. anyway, good to know.

**Comment by JCBrouwer** [comment]
Ok I took a look at it yesterday and pushed it [here](https://github.com/maua-maua-maua/maua/commit/60fd560c70ba84ac1ac4eba19a23fa677f988437). There's still a few bugs in the optical flow encoding and caching which I need to iron out, but this should be a good upgrade over the NpyAppendArray implementation. It's quite a bit faster on my system even for long videos and with the frames encoded as JPGs on disk.

**Comment by JCBrouwer** [comment]
Ok going to close this as I think all the slow-down issues are fixed with the new version I've merged.

I'm tracking the newly introduced artifacts in #7 

Feel free to open up if the issue persists on your end
